packages/common/src/ContractInteractor.ts

```diff
@@ -397,16 +401,85 @@ export class ContractInteractor {
   }
 
   async getPastEventsForHub (extraTopics: string[], options: PastEventOptions, names: EventName[] = ActiveManagerEvents): Promise<EventData[]> {
-    return await this._getPastEvents(this.relayHubInstance.contract, names, extraTopics, options)
+    return await this._getPastEventsPaginated(this.relayHubInstance.contract, names, extraTopics, options)
   }
 
   async getPastEventsForStakeManager (names: EventName[], extraTopics: string[], options: PastEventOptions): Promise<EventData[]> {
     const stakeManager = await this.stakeManagerInstance
-    return await this._getPastEvents(stakeManager.contract, names, extraTopics, options)
+    return await this._getPastEventsPaginated(stakeManager.contract, names, extraTopics, options)
   }
 
   async getPastEventsForPenalizer (names: EventName[], extraTopics: string[], options: PastEventOptions): Promise<EventData[]> {
-    return await this._getPastEvents(this.penalizerInstance.contract, names, extraTopics, options)
+    return await this._getPastEventsPaginated(this.penalizerInstance.contract, names, extraTopics, options)
+  }
+
+  async getPagesForBlockWindow (fromBlock: BlockNumber = 1, toBlock?: BlockNumber): Promise<number> {
+    // save 'getBlockNumber' roundtrip for a known max value
+    if (this.maxPageSize === Number.MAX_SAFE_INTEGER) {
+      return 1
+    }
+    if (toBlock === 'latest' || toBlock == null) {
+      toBlock = await this.getBlockNumber()
+    }
+    // noinspection SuspiciousTypeOfGuard - known false positive
+    if (typeof fromBlock !== 'number' || typeof toBlock !== 'number') {
+      throw new Error('ContractInteractor:getPartsForBlockWindow: only number supported for block range when using pagination')
+    }
+    const rangeSize = toBlock - fromBlock + 1
+    const pagesForBlockWindow = Math.max(Math.ceil(rangeSize / this.maxPageSize), 1)
+    this.logger.info(`Splitting request for ${rangeSize} blocks into ${pagesForBlockWindow} smaller paginated requests!`)
+    return pagesForBlockWindow
+  }
+
+  splitRange (fromBlock: BlockNumber, toBlock: BlockNumber, parts: number): Array<{ fromBlock: BlockNumber, toBlock: BlockNumber }> {
```

Comment:
This function is too similar to `getPagesForBlockWindow`, and they have the same calculation inside - which is generally a bad thing. In this case, you actually have a bug that if you provide to any `getPastEvents` function in interactor `toBlock = 'latest'`, it will throw in line 441, even though `getPagesForBlockWindow` supports it.

Reply:
They do have a same calculation but do the opposite things. `splitRange` has existed before this change and is responsible for translation of 1 range into X smaller ranges. `getPagesForBlockWindow` determines X. The common code is boilerplate and can be extracted, generally speaking.